### PR TITLE
nodejs-16_x: 16.5.0 -> 16.6.1

### PR DESCRIPTION
**Pull request by piemadd**
###### Motivation for this change
https://github.com/nodejs/node/releases/tag/v16.6.1
TLDR:
- Brings NPM version to 7.20.3
- Fixes some bugs
- (Hopefully) Fixes issue with building for Darwin (16.6.0 wouldn't build for some reason - https://github.com/NixOS/nixpkgs/commit/c73d1006a216134713593a54686bca21976b443e)

> BTW this is one of my first commits to a large repo so if I've done anything wrong please let me know so I can fix it. I've done my best to follow examples and instructions, though I know that due to me being new I am prone to mistakes. Thank you.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 133214` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>lumo</li>
    <li>nodejs-16_x (nodejs_latest)</li>
    <li>nodejs-slim-16_x (nodejs-slim_latest)</li>
  </ul>
</details>

**Comment by happysalada** [comment]
Some more details.
The problem with node 16.6 is that it introduces features not available in the sdks prior to 11 (which is the default for macos).

Somebody pointed to the commit where this was introduced in v8. There could be a chance of reverting that commit with a patch. I just haven't gotten to it yet.

**Comment by pieromqwerty** [comment]
That makes sense, I've had some similar issues with Python on Ubuntu-based distros due to 3.6 being the default, so I see how that could happen. Seems like the fail in the darwin tests was from a failed network connection though, not anything programmatic, weird.

**Comment by happysalada** [comment]
The logs are not very verbose. 
There are two problems
```
WARNING: C++ compiler (CXX=clang++, 7.1.0) too old, need g++ 8.3.0 or clang++ 8.0.0
```
There is an effort at the moment to update the llvm to version 11 but that might still take some time.

The other problem didn't show, but it's here in the code
https://github.com/nodejs/node/blob/e83c8aef4b787b531f23ad7abb498fe81db5ee83/deps/v8/src/libplatform/tracing/recorder.h#L17
this header is not available on apple_sdk prior to version 11 I think. So this will probably still fail.

Thank you for the PR though!

**Comment by pieromqwerty** [comment]
Ooh that makes more sense, though it does make me wonder how 16.5.0 built properly 🤔.

**Comment by pieromqwerty** [comment]
Ooh that makes sense. I assume in this case the best course of action would be instead to use a patch instead of changing the version altogether?

**Comment by happysalada** [comment]
yes I would think it would be the easiest to merge. I'm not sure how much of a pain it would be though.

**Comment by pieromqwerty** [comment]
Seems like it was the case for getting a 16.6.0 patch already in, shouldn't be _too_ hard though, I already have the base32 hash for that `.tar.xz`.

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 133214` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>lumo</li>
    <li>nodejs-16_x</li>
    <li>nodejs-slim-16_x</li>
  </ul>
</details>

**Comment by happysalada** [comment]
Here is what the log says
```
error: hash mismatch in fixed-output derivation '/nix/store/h1nhmnj5iaprvwid7lh93v9wcdvg9d7v-3d53ff8ff0e721f908d8aff7a3709bc6dbb07ebb.patch.drv':
         specified: sha256-ebHqBYzGfyppRizV8kZ6Hv4IxkKZwFPacDhM4aDj5Vc=
            got:    sha256-EXz1QMyQ9jMXadCIhrFruqgREKrhjEMNOVYkN95ao7w=
```

**Comment by pieromqwerty** [comment]
Hm weird, I guess I used the wrong hash then.
